icu4c/source/common/loclikely.cpp

```diff
@@ -1241,12 +1239,16 @@ static UBool
 _ulocimp_addLikelySubtags(const char* localeID,
                           icu::ByteSink& sink,
                           UErrorCode* status) {
-    char localeBuffer[ULOC_FULLNAME_CAPACITY];
-
-    if (do_canonicalize(localeID, localeBuffer, sizeof localeBuffer, status)) {
-        return _uloc_addLikelySubtags(localeBuffer, sink, status);
+    PreflightingLocaleIDBuffer localeBuffer;
+    do {
+        *(localeBuffer.getCapacityPtr()) = do_canonicalize(localeID, localeBuffer.getBuffer(), localeBuffer.getCapacity(), status);
```

Comment:
Re: `*(localeBuffer.getCapacityPtr()) = `

FWIW, this seems a bit odd (at least to me) -- assigning the resulting size to the "capacity".

Ex: For the stack-allocated case where the data fits, it would mean that the `PreflightingLocaleIDBuffer` now thinks the capacity is *less* than the actual buffer length (`ULOC_FULLNAME_CAPACITY`).

It would also mean that the new capacity wouldn't include the null terminator, as the resulting size reported by `uloc_canonicalize` doesn't include the terminating null in the size. 
So for something like "en-US", it would think the capacity was only 5.

On the other hand, when the data *doesn't fit* then the capacity would be the same as the data size + 2.

Maybe I'm just getting confused by the term "capacity" though?
I wonder if perhaps calling it "size" instead of "capacity" would be less confusing?

Reply:
I decided to split it into two variables: `capacity`, which gives the current capacity of the buffer, and `requestedCapacity`, which is the capacity that the call we're wrapping says it needs.  I agree this is clearer, but it comes at the expense of using two variables instead of one, since the real capacity and the requested capacity are never used at the same time if you use this class the way it was intended to be used.  I could just change the name of that setter function to make it clearer what the caller is doing and leave things with a single internal variable (maybe under a new name), but the extra four (or eight) bytes aren't that big a deal and it _is_ clearer this way, I think...

By the way, I acknowledge that `*(buffer.getRequestedCapacityPtr()) = uloc_doSomething(...);` looks a little weird.  What I was trying to do was stick as close as I could to the way you'd normally use a function like this: `bytesNeeded = uloc_doSomething(...);`.  I was concerned that using a regular setter function would make it harder to see what the calling code was actually doing: `buffer.setRequestedCapacity(uloc_doSomething(...));` kind of buries the `uloc_doSomething(...)` part, and that's really where the reader's focus should be.  If you guys think what I'm doing actually hurts readability more than a normal setter function (or have some other better idea), let me have it.

Reply:
Thanks for explaining about the `*(buffer.getRequestedCapacityPtr()) = uloc_doSomething(...);` style.
FWIW, I think that the approach with the `do... while` loop helps to make it more clear, as it puts the focus on the method being called, rather than the logic of allocating the buffer, etc..

Regarding splitting the variables, I think that helps to make it more clear, but maybe Frank/others might feel otherwise? I'm not sure what their thoughts might be.

Though one comment/issue: with them split like this now, I wonder if we might end up with odd cases in error situations...

In `needToTryAgain`, if the memory allocation fails, since we update the value of `capacity` first, we can have a case where  `capacity` gets set to a larger value, but the `heapBuffer` isn't allocated. I'm not sure how much it matters though, as the `needToTryAgain` will return FALSE, but it might be curious to have some `PreflightingLocaleIDBuffer` objects with both `requestedCapacity` and `capacity` larger than `ULOC_FULLNAME_CAPACITY`, yet no actual heap buffer.

I wonder if we might want to do something like this perhaps?
```c++
    newCapacity = requestedCapacity + 2;    // one for the terminating null, one just for paranoia
    heapBuffer = static_cast<char*>(uprv_malloc(newCapacity));
    if (heapBuffer == nullptr) {
        *err = U_MEMORY_ALLOCATION_ERROR;
    } else {
        *err = U_ZERO_ERROR;
        capacity = newCapacity;
    }
    return U_SUCCESS(*err);
```

Reply:
>I wonder if we might want to do something like this perhaps?

Sure...